chrome/src/browser/index.ts

```diff
@@ -0,0 +1,125 @@
+/**
+ * The entry point for the Chrome app.
+ */
+import {smarthomeStub, AppStub} from 'local-home-testing';
+import {PlatformCoordinator} from './platform-coordinator';
+import {UDPScanConfig} from 'local-home-testing/build/src/radio/dataflow';
+
+/**
+ * Initialize the UI DOM.
+ * Redefine all `console.log` and `console.error` calls to log
+ * to an element in the page.
+ */
+const logElement = document.getElementById('log');
+if (logElement) {
+  const modifiedConsole: any = console;
+  modifiedConsole.defaultlog = console.log;
+  modifiedConsole.log = (message: string) => {
+    modifiedConsole.defaultlog(message);
+    if (logElement) {
+      logElement.innerHTML += `[${new Date().toLocaleTimeString()}]:<br/>${message}<br/>`;
+      // Scrolls to the bottom of the element.
+      logElement.scrollTop = logElement.scrollHeight;
+    }
+  };
+  modifiedConsole.error = modifiedConsole.log;
+}
+
+/**
+ * Callback to hook into app initialization.
+ */
+let onAppInitialized: (app: AppStub) => void | undefined;
+
+/**
+ * A promise to resolve when the app constructor gets called.
+ */
+const saveAppStub = new Promise<AppStub>(resolve => {
+  onAppInitialized = (app: AppStub) => {
+    resolve(app);
+  };
+});
+
+/**
+ * Override the constructor to access the stub on creation.
+ */
+class ChromeAppStub extends AppStub {
+  constructor(version: string) {
+    super(version);
+    onAppInitialized(this);
+  }
+}
+
+/**
+ * Set the global dependencies.
+ */
+smarthomeStub.App = ChromeAppStub;
+export const smarthome = smarthomeStub;
+
+/**
+ * Listen for the app bundle file picker.
+ */
+const filePicker = document.getElementById('app-file-picker');
+if (filePicker) {
+  filePicker.addEventListener('change', async event => {
+    const files = (event.target as HTMLInputElement).files;
+    if (files) {
+      // Run the javascript.
+      eval(await files[0].text());
```

Comment:
should you run that in a worker?

Reply:
I don't have a strong opinion whether we should or not.  This should happen almost instantly, and we don't need a reference to the app after we get the handlers.  On the other hand the `eval` call seems a bit out of place here.  Maybe @Fleker would have an opinion.

Reply:
Using a worker is probably the preferred method just for performance, and `eval` is typically not advised.

Reply:
Hmm, the different global contexts is making it a bit tricky.

I can't pass the stubs between the worker thread and the main thread, since it appears objects have to be [`Transferable`](https://developer.mozilla.org/en-US/docs/Web/API/Transferable) to be sent to/from a `WebWorker`.  This makes it difficult to set the handlers from a worker thread.

The best solution would probably be to run the `PlatformCoordinator` entirely in a worker thread, and have `index.ts` strictly handle UI.  Then parameters to the worker in messages, and have the worker respond to the messages with results.  This would be similar to how `CLI` works now.  However we'd be adding another layer to the app and would need to implement message handling, so I'm not sure it's worth for the performance boost.

Also, looking around for an alternative to `eval`, I'm not sure there are many ideal alternatives.  As I understand the main concern with `eval` is that it's running arbitrary javascript which isn't exactly safe.  There's the [FileSystem API](https://developer.mozilla.org/en-US/docs/Web/API/FileSystem), which might(?) let us access the javascript file directly rather than loading the text, but inherently we're still loading arbitrary javascript.

I can open issues for these if needed (for discussion as well).  Not sure if there is a simple solution I'm missing.